_P251_PZEM004Tv3.ino

```diff
@@ -0,0 +1,394 @@
+
+        // Sequence for changing PZEM address
+        if (P251_PZEM_ADDR_SET == 1)    //if address programming confirmed
+        {
+          P251_PZEM_sensor->setAddress(P251_PZEM_ADDR);
+          P251_PZEM_mode = 0;       // Back to read mode
+          P251_PZEM_ADDR_SET = 3;   // Address programmed
+        }
+      }
+        P251_PZEM_sensor->init(P251_PZEM_ADDR);
+
+        // Sequence for reseting PZEM energy
+        if (P251_PZEM_mode == 1)
+        {
+          P251_PZEM_sensor->resetEnergy();
+          P251_PZEM_mode = 0;       // Back to read mode
+          P251_PZEM_ADDR_SET = 4;   // Energy reset done
+        }
+
+        Plugin_251_init = true;
+        success = true;
+        break;
+      }
+
+    case PLUGIN_EXIT:
+      {
+        if (P251_PZEM_FIRST == event->TaskIndex)  //If first PZEM, serial config available
+        {
+          if (P251_PZEM_sensor)
+          {
+            delete P251_PZEM_sensor;
+            P251_PZEM_sensor=nullptr;
+          }
+        }
+        break;
+      }
+
+    case PLUGIN_READ:
+      {
+        if (Plugin_251_init && P251_PZEM_mode==0 )    //Read sensor
+        {
+          // When new data is available, return true
+          P251_PZEM_sensor->init(P251_PZEM_ADDR);
+
+          float PZEM[6];  
+          PZEM[0] = Plugin251_ReadVoltage();
+          PZEM[1] = Plugin251_ReadCurrent();
+          PZEM[2] = Plugin251_ReadPower();
+          PZEM[3] = Plugin251_ReadEnergy();
+          PZEM[4] = Plugin251_ReadPf();
+          PZEM[5] = Plugin251_ReadFreq();
+             
+          UserVar[event->BaseVarIndex]     = PZEM[P251_QUERY1];
+          UserVar[event->BaseVarIndex + 1] = PZEM[P251_QUERY2];
+          UserVar[event->BaseVarIndex + 2] = PZEM[P251_QUERY3];
+          UserVar[event->BaseVarIndex + 3] = PZEM[P251_QUERY4];
+          sendData(event);   //Permet d'envoyer à l'exterieur du plugin (vers controller ou vers trigger de rules)
+          success = true;
+        }
+        break;
+      }
+
+     case PLUGIN_WRITE:
+    {
+      if (Plugin_251_init)
+      {
+        String command = parseString(string, 1);
+        if (command == F("resetenergy") && (P251_PZEM_FIRST == event->TaskIndex))
+        {
+          if ((event->Par1 >= 0) && (event->Par1 <= 247))
+          {
+            P251_PZEM_sensor->init(event->Par1);
+            P251_PZEM_sensor->resetEnergy();
+            success = true;
+          }
+        }
+      }
+
+       break;
+    }
+  }
+  return success;
+}
+
+//************************************//
+//***** reading values functions *****//
+//************************************//
+
+// NOTE: readings are attempted only PZEM_AMX_ATTEMPT times
+
+float Plugin251_ReadVoltage() {
+  int counter = 0;
+	float reading = -1.0;
+	do {
+		reading = P251_PZEM_sensor->voltage();
+		wdt_reset();
+		counter++;
+	} while (counter < PZEM_MAX_ATTEMPT && reading < 0.0);
+  if (reading == -1) reading = 0;
+	return reading;
+}

```

Comment:
These `Plugin251_ReadXXX` functions are copy/paste and I think you can make it a lot more readable by just using a switch statement and a parameter for the function to state what you're reading.

Also these functions are now blocking, which is also shown by calls to the `wdt_reset()` function in each of them.
Please just try to read them once and if it needs another attempt try again in a next loop.
On a multi-sensor system you must not have blocking calls like these.

Reply:
I finally remove these functions. I check validity of received data in the PLUGIN_READ.
I have an issue that I don't understand how to solve: I need to include PZEM004Tv30.h directly in ESPEasy.ino otherwise I get this error "ESPEasy_mega-20191003/source/src/ESPEasy.ino:1526:29: error: 'PZEM004Tv30' was not declared in this scope". How to solve this ?